ddtrace/internal/datadog/profiling/stack_v2/src/sampler.cpp

```diff
@@ -54,13 +55,94 @@ create_thread_with_stack(size_t stack_size, Sampler* sampler, uint64_t seq_num)
+    auto new_process_count = static_cast<uint64_t>(ts.tv_sec * 1000000ULL + ts.tv_nsec / 1000);
+
+    clock_gettime(CLOCK_THREAD_CPUTIME_ID, &ts);
+    auto new_sampler_thread_count = static_cast<uint64_t>(ts.tv_sec * 1000000ULL + ts.tv_nsec / 1000);
+#elif defined(__MACH__)
+    // Get the process CPU time
+    task_thread_times_info_data_t task_info_data;
+    mach_msg_type_number_t task_info_count = TASK_THREAD_TIMES_INFO_COUNT;
+
+    if (task_info(
+          mach_task_self(), TASK_THREAD_TIMES_INFO, reinterpret_cast<task_info_t>(&task_info_data), &task_info_count) !=
+        KERN_SUCCESS) {
+        return;
+    }
+
+    auto new_process_count =
+      static_cast<uint64_t>(task_info_data.user_time.seconds * 1e6 + task_info_data.user_time.microseconds +
+                            task_info_data.system_time.seconds * 1e6 + task_info_data.system_time.microseconds);
+
+    // Get the sampling thread CPU time
+    mach_msg_type_number_t count = THREAD_BASIC_INFO_COUNT;
+    thread_basic_info_data_t info;
+
+    thread_port_t thread = mach_thread_self(); // perf: call once
+    int kr = thread_info(thread, THREAD_BASIC_INFO, reinterpret_cast<thread_info_t>(&info), &count);
+    mach_port_deallocate(mach_task_self(), thread);
+    if (kr != KERN_SUCCESS) {
+        return;
+    }
+
+    auto new_sampler_thread_count =
+      static_cast<uint64_t>(info.user_time.seconds * 1e6 + info.user_time.microseconds +
+                            info.system_time.seconds * 1e6 + info.system_time.microseconds);
+#endif
+    auto sampler_thread_delta = static_cast<double>(new_sampler_thread_count - sampler_thread_count);
+    auto process_delta = static_cast<double>(new_process_count - process_count - sampler_thread_delta);
+    if (process_delta <= 0) {
+        process_delta = 1; // Avoid division by zero or negative values
+    }
+
+    auto current_interval = static_cast<double>(sample_interval_us.load());
+
+    // We assume that every sampling operation contributes a fixed amount of
+    // overhead, while the application consumes an average amount of CPU over
+    // time. With:
+    //    s - sampler time
+    //    p - process time
+    //    o - overhead threshold
+    //    I - interval
+    //    I'- interval after adjustment
+    // we use the following formula to adapt the sampling interval
+    //    I' = I * [(s / p) / o]
+    // As the value could be small when the process is idle, we use a lower
+    // bound of the sampling interval to avoid CPU spikes from the sampler.
+    auto new_interval =
+      static_cast<microsecond_t>(current_interval * ((sampler_thread_delta / process_delta) / g_target_overhead));
+
+    // Cap the new interval to the min/max sampling period
+    if (new_interval < g_min_sampling_period_us) {
+        new_interval = g_min_sampling_period_us;
+    } else if (new_interval > g_max_sampling_period_us) {
+        new_interval = g_max_sampling_period_us;
+    }
+
+    sample_interval_us.store(new_interval);
```

Comment:
I wonder if we still need this to be atomic since we probably only set the interval from just one thread now 🤔

Reply:
I wouldn't mind dropping `atomic` for this if you can remove all invocations of `set_interval()` from the code. There are a few. 

https://github.com/DataDog/dd-trace-py/blob/8613e0808e07d1e746a85cf7cf9f08c229973330/ddtrace/internal/datadog/profiling/stack_v2/src/stack_v2.cpp#L23

Reply:
👍  we can look into it in a follow-up